### DISCUSSION TITLE
IRL FPGA communication

### DISCUSSION THREAD

**Discussion by Praetrorian** [cocotb/cocotb#3677] [Q&A] [answered]
Hello all,

In my company, we are exploring ways to replace our home-framework of simulation (with Modelsim) for multiple reasons.

Cocotb sounds promising as it interfaces with Modelsim and Verilator (which we start using 2 years ago).

However, the world of open-source is what it is and finding the information I need is a challenge itself.

I have a question regarding the communication with a real FPGA. On the main page of cocotb.org : 
`"Do you want to talk to a golden model in your testbench? Or to real hardware, e.g. an FPGA or a logic analyzer? In most cases, that's just a matter of looking for existing Python bindings"`

This communication with real hardware could make us do the leap towards Cocotb as even with Verilator our simulations are crazy long.
But I couldn't find any example, any developer doing that (maybe I don't have the good keywords).

Do you guys have any reference/doc about it?

Thank you

**Comment by ktbarrett** [comment, marked as answer]
## Existing projects

Look at [basil](https://github.com/SiLab-Bonn/basil). @themperek could tell you more about this.

## Documentation

> Do you guys have any reference/doc about it?

This is going to be highly specific depending upon what you are testing, what your requirements are, etc. This is more something that would be helped by examples or review articles. Unfortunately I think the issue is most of the people doing this right now are doing it behind closed doors, and making those examples or analyses public is not possible.

Personally, I can say we did invest in this at a previous company. We accomplished it by creating a compatibility layer that was used to write cocotb-esque tests that dispatched to one of two "backend" (simulation or HW). The HW backend provided cocotb functionality by leveraging asyncio and pytest, and we put in effort to provide the synchronization primitives and testing functionality that cocotb provides.

An example that is *not* exactly what we did...
```python
import our_framework as of
from my_design import Testbench

@of.test(tb=Testbench, ...)  # accepts same args as cocotb.test
async def test_thing(tb: Testbench):  # must start with `test` for pytest, you get a testbench object as `dut` object is impossible in HW

    # models behavior so we can check expected outputs, this could be written in Python, C/C++ with pybind11 binding,
    # or MATLAB which was what our prototypers/scientists wrote in and would throw over the wall to us
    model = FWModel(...)

    async def do_writes():
        for _ in range(TEST_LENGTH):
            data = randint(1, 100)
            await tb.write(data)
            model.send(data)

    async def check_output():
        for _ in range(TEST_LENGTH):
            data = await tb.read()
            assert (await model.recv()) == data

    # join_all is Combine in cocotb, gather() in asyncio for HW testing, we had to provide this compatibility later
    await join_all(do_writes(), check_output())
```

Maybe there is a way to create a generic cocotb-like asyncio/pytest-backend framework, but no *public* work has been done on that yet. 

## Is this even a good idea?

For device-level or system-level testing while I think the above is a sufficient idea, but has some downsides.
1. Test duplication: It's likely there will be C/C++ integration testing done and any tactical software code will be done using C or C++, so invariably you end up writing the same test twice. It just might be on two different teams. This encourages "siloing" rather than cooperation.
2. Only supporting cocotb/HW compat framework tests for device/system-level testing means you will end up rewriting failing integration tests in that framework so you can debug. This is not always straightforward or feasible ("I rewrote it and the bug isn't present anymore!").

I think there are better ideas. For example, at the level of device/system verification it is better to write C or C++ tests that can be run in simulation or on HW.
1. Tests written by the SW team, the FW team, and manufacturing (integration and qualification testing team) can be re-used by *all* the others. This encourages cooperation.
2. The tests can test C/C++ code in HDL simulation; meaning you can see bugs in the waveform viewer and can test SW drivers during development.
3. The software team and the integration team know how to write C/C++. By using a cocotb/HW compat framework in this role, you are forcing them to learn the company-specific cocotb/HW compat framework.

One way I've heard people do this (though I have never tried it personally) is by using qemu to emulate the processor, hooking memory accesses via qemu's builtin solution, and doing RPC into a simulator via either VPI/VHPI/FLI calls or via TLM. This shouldn't be terribly difficult to set up and is *extremely* flexible.

## Conclusion

cocotb has a time and place and while it is possible to write HW tests using cocotb, I don't think it's the best course of action. I would use cocotb for IP core development and some device/system-level IP integration, while introducing a C/C++ framework to the majority of device and system-level HW verification and integration testing. This is how I've been pushing my current employer, and how it was in process of being set up at my previous employer.

**Reply by Praetrorian** [reply]
Thank you for this very detailed answer, much appreciated.

We develop space equipments that often require us to do simulation that last multiple seconds (days/week of computer time).

These equipments are tested with EGSE developed by a different team and they use their own framework. I don't think we could be able to develop a single test usable by both team (and in our domain we like the fact that different test are done for robustness). Mainly because the FPGAs are not tested themselves by the hw team, but the whole equipment is.

However, I'm glad to hear that it is doable, even if we might have to develop a dedicated PCB.

Thank you again for your answer, it will help us to make our decision.

**Reply by marlonjames** [reply]
At a previous employer, I developed an automated testing system with cocotb for testing FPGAs.
We wrote the requirements-based tests in cocotb, collected coverage, etc, in simulation, then used those simulations to create stimulus that we fed into the FPGA. FPGA outputs were captured during the test. Then we fed the stimulus and captured outputs back into a "fake" simulator with cocotb on top, and re-ran the tests to verify they passed.
There are some caveats, such as making your BFMs and tests tolerant of the discrete nature of sampled output capture vs perfect simulation, and you may need to skew the captures to account for PCB trace length differences, clock drift etc, but it worked well and significantly shortened the cycle of perform test -> Engineers verify test results.

**Comment by themperek** [comment]
You can look here:  https://github.com/themperek/ORConf23/ and slides [here](https://github.com/themperek/ORConf23/blob/master/ORConf23-Reusable.pdf)

or something different:  https://github.com/cocotb/cocotb/discussions/3615#discussioncomment-8155841

one could do a "fake" simulator module that toggles physical pins ...
